### PR TITLE
MM-37294 MM-37296 MM-37297 MM-37495 MM-37509 CRT fixes

### PR DESCRIPTION
**Pull request by anurag6713**
#### Summary
* Fixes "theme not updated on switching the team" issue
* Users can follow posts with no replies
* Fixed unable to unfollow post with no replies
* Fixed warnings when rendering post options when CRT is not enabled
* Enables MARK ALL AS READ even when there are no unreads to temporarily fix unreads count issue.
* Fix auto following threads on UI when opening a thread. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-37294
https://mattermost.atlassian.net/browse/MM-37296
https://mattermost.atlassian.net/browse/MM-37297
https://mattermost.atlassian.net/browse/MM-37495
https://mattermost.atlassian.net/browse/MM-37509

#### Device Information
This PR was tested on: iOS 14.5, iPhone 12 Simulator

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: app/components/global_threads/thread_footer/thread_footer.tsx [hunk 1]
```diff
@@ -132,18 +132,21 @@ function ThreadFooter({actions, currentUserId, intl, location, testID, theme, th
...
-                return [];
+        if (thread.participants?.length) {
+            let isThreadStarterFound = false;
+            const participantIds = thread.participants.flatMap((participant) => {
+                if (participant.id === threadStarter?.id) {
+                    isThreadStarterFound = true;
+                    return [];
+                }
+                return participant.id;
+            });
```
Comment (koox00) [inline app/components/global_threads/thread_footer/thread_footer.tsx:143]: oh cool! didn't know about that trick, using `flatMap` to `map` and `filter` at the same time!

File: app/mm-redux/selectors/entities/threads.ts [hunk 2]
```diff
@@ -52,7 +52,7 @@ export function getThread(state: GlobalState, threadId: $ID<UserThread>, fallbac
     if (!thread || !thread?.id) {
         if (fallbackFromPosts) {
             const post = getPost(state, threadId);
-            if (post?.participants?.length) {
+            if (post && !post?.root_id) {
```
Comment (koox00) [inline app/mm-redux/selectors/entities/threads.ts:55]: nit: the `?` is redundant in this case
```suggestion
            if (post && !post.root_id) {
```

**Review by koox00 — APPROVED** [review]
LGTM!

**Review by josephbaylon — APPROVED** [review]
LGTM. The warning bubble is now gone when post options is opened.

**Comment by jgilliam17** [comment]
Thanks @anurag6713 
Theming when switching teams looks good and following/unfollowing single post is working as expected.
One small text change - long press menu for single post(not a thread yet) should match webapp with  "Follow message".

File: app/components/global_threads/thread_list/thread_list_header/index.tsx [hunk 3]
```diff
@@ -62,13 +62,13 @@ function ThreadListHeader({haveUnreads, intl, markAllAsRead, style, testID, view
             </View>
             <View style={style.markAllReadIconContainer}>
                 <TouchableOpacity
-                    disabled={!haveUnreads}
+                    disabled={false/*!haveUnreads*/}
```
Comment (enahum) [inline app/components/global_threads/thread_list/thread_list_header/index.tsx:65]: Why? Is then the prop not really needed?

File: app/screens/post_options/post_options.js [hunk 4]
```diff
@@ -110,19 +111,27 @@ export default class PostOptions extends PureComponent {
     }
 
     getFollowThreadOption = () => {
-        const {thread} = this.props;
-        if (!thread) {
+        const {location, thread} = this.props;
+        if (location !== CHANNEL) {
```
Comment (enahum) [inline app/screens/post_options/post_options.js:115]: So cannot follow a thread from the channel screen?

File: app/screens/post_options/post_options.js [hunk 5]
```diff
@@ -110,19 +111,27 @@ export default class PostOptions extends PureComponent {
...
         let message;
-        if (thread.is_following) {
+        if (thread?.is_following) {
             icon = 'message-minus-outline';
-            message = {id: t('threads.unfollowThread'), defaultMessage: 'Unfollow Thread'};
+            if (thread?.participants?.length) {
+                message = {id: t('threads.unfollowThread'), defaultMessage: 'Unfollow Thread'};
+            } else {
+                message = {id: t('threads.unfollowMessage'), defaultMessage: 'Unfollow Message'};
+            }
```
Comment (enahum) [inline app/screens/post_options/post_options.js:127]: No need for if/else as they both are doing exactly the same

File: app/screens/post_options/post_options.js [hunk 6]
```diff
@@ -110,19 +111,27 @@ export default class PostOptions extends PureComponent {
...
         } else {
             icon = 'message-plus-outline';
-            message = {id: t('threads.followThread'), defaultMessage: 'Follow Thread'};
+            if (thread?.participants?.length) {
+                message = {id: t('threads.followThread'), defaultMessage: 'Follow Thread'};
+            } else {
+                message = {id: t('threads.followMessage'), defaultMessage: 'Follow Message'};
```
Comment (enahum) [inline app/screens/post_options/post_options.js:133]: Huh?

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/post_options/post_options.js (hunk 6 above)
Comment (anurag6713) [reply app/screens/post_options/post_options.js:133]: Follow/Unfollow "THREAD" or "MESSAGE" ..

File: app/screens/post_options/post_options.js (hunk 4 above)
Comment (anurag6713) [reply app/screens/post_options/post_options.js:115]: You can follow/unfollow a thread/message by long pressing the posts on the channel screen.

File: app/components/global_threads/thread_list/thread_list_header/index.tsx (hunk 3 above)
Comment (anurag6713) [reply app/components/global_threads/thread_list/thread_list_header/index.tsx:65]: https://mattermost.atlassian.net/browse/MM-37495

As we are looking for a temporary solution, thought commenting would be a quick fix.

**Review by enahum — APPROVED** [review]

**Comment by anurag6713** [comment]
@jgilliam17 I have updated the text accordingly and also included this ticket as well https://mattermost.atlassian.net/browse/MM-37495.

File: app/components/sidebars/main/channels_list/list/list.js [hunk 7]
```diff
@@ -400,8 +400,7 @@ export default class List extends PureComponent {
     };
 
     render() {
-        const collapsedThreadsEnabled = true;
-        const {testID, styles, theme} = this.props;
+        const {collapsedThreadsEnabled, styles, testID, theme} = this.props;
```
Comment (enahum) [inline app/components/sidebars/main/channels_list/list/list.js:403]: Thank you!

**Review by enahum — APPROVED** [review]
Let's aim to merge & cherry-pick as fast as possible

**Comment by enahum** [comment]
@anurag6713 test failing

**Review by enahum — APPROVED** [review]

File: app/actions/websocket/threads.ts [hunk 8]
```diff
@@ -25,7 +25,9 @@ export function handleThreadReadChanged(msg: WebSocketMessage) {
         if (msg.data.thread_id) {
             const state = getState();
             const thread = getThread(state, msg.data.thread_id);
-            if (thread) {
+
+            // Mark only following threads as read.
+            if (thread?.is_following) {
```
Comment (hahmadia) [inline app/actions/websocket/threads.ts:30]: Thanks Anurag ❤️

**Review by jgilliam17 — APPROVED** [review]
Thanks @anurag6713 
Tested, looks good to merge.
- Verified theming when switching teams, following/unfollowing single post, mark as read is always enabled and ‘follow message’ text on single post (not yet a thread).
Also, closing/opening the app and opening an unfollowed thread no longer auto-follows that thread.
.

**Pull request merged**